### ISSUE TITLE
Airflow DAG to ping pull request reviewers

### ISSUE THREAD

**Issue by zackkrida** [WordPress/openverse-catalog#552]
## Problem

We have a lot of pull requests that go stale; it would be nice to not have to manually ping reviewers for pull requests via Slack DM or `@` in GitHub.

## Description

Write a script + DAG to schedule the script that can automatically ping PR reviewers.

## Additional context

This work will require some new folder structure in this repo for DAGs.

## Implementation

- [ ] 🙋 I would be interested in implementing this feature.

**Comment by zackkrida** [comment]
You could pin PyGithub to [1.54.1](https://pygithub.readthedocs.io/en/latest/changes.html#version-1-54-1-december-24-2020) but that's probably not the best idea.

**Comment by sarayourfriend** [comment]
That version is pretty old at this point :cry: Luckily we really only need to make a few targeted requests:

1. Get list of open PRs
2. Get comments on PRs
3. Post comments on PRs

I'm hopeful that this won't be too terribly cumbersome.

**Comment by AetherUnbound** [comment]
RE PyGitHub, Airflow dependencies can be a complicated mess, so we use a [constraints file](https://airflow.apache.org/docs/apache-airflow/stable/installation/installing-from-pypi.html#constraints-files) to lock in dependencies. It does look like PyGithub is currently pinned to 1.54.1 as Zack mentioned. That's _probably_ fine, but an `HttpHook` would also be totally reasonable.

As for the "dagification", go with what is easiest! Using a `PythonOperator` with the `HttpHook` is a totally idiomatic way of writing a DAG 😄 You _could_ split that up into separate functions, passing parameters between them using XComs if you'd like. Or it could all be just one operator. Since this is a pretty straightforward DAG, you could also try using the [TaskFlow API](https://airflow.apache.org/docs/apache-airflow/stable/tutorial_taskflow_api.html)! I haven't used it yet, but I've been wanting to for some time, might be the perfect opportunity.

**Comment by sarayourfriend** [comment]
Okay. I've written it as a single function to pass to `PythonCallable`. It uses `HttpHook`, but the secret is pulled in the DAG definition file using `Variable` and passed to the callable with `op_args`. Is that okay or does it need to happen another way?

I'm not seeing the benefit of the TaskFlow API in this case. It's a single function, there isn't really a "data" pipeline :thinking: We're not saving any of the data as I think it's wiser to just treat GitHub itself as the source of truth.

**Comment by sarayourfriend** [comment]
@AetherUnbound and @rwidom (I remember y'all working on this) I need some help with the URI encoded connection strings. For the GitHub connection I need to be able to pass headers as a query param to the final URL encoded string, but it seems that's not possible with the current "automatic url encoding" script :thinking: Or in any case, I'm having a heck of a time trying to get it to work with the automatic encoding. Looking at the example at the bottom of this page, for example, I'm unable to get it to work: https://airflow.apache.org/docs/apache-airflow-providers-http/stable/connections/http.html

Instead of using the headers param it treats the connection string as the base url and I end up making requests that look like this:

```
[2022-06-10, 10:46:03 UTC] {http.py:129} INFO - Sending 'GET' to url: https://api.github.com?headers=%7B'Authorization':'token%20<redacted>'%7D/repos/WordPress/openverse/pulls
```

As you can see, that's not a properly formed URI!

Here's the template I'm using for the connection string:

```
AIRFLOW_CONN_GITHUB="https://api.github.com?headers=%7B'Authorization':'token%20%3Cnot_set%3E'%7D"
```

Obviously with `not_set` replaced by an actual PAT in my .env file.

**Comment by AetherUnbound** [comment]
> @AetherUnbound something I'm unclear about with our "deployless Airflow" setup that we have now with the DAG pull job and all... how do we get new variables into the production Airflow? Do we define them through the Airflow admin instead of the .env file? Or is there a way to pull new environment variables in and have Airflow pick them up?

For now, we're adding them through the UI. In order to pull them from environment variables, we'd either need to 1) redeploy airflow entirely or 2) leverage the [airflow CLI](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#connections) during the DAG sync process to update the connections/variables automatically. More discussion on this here: https://github.com/WordPress/openverse-catalog/issues/209. Down the road we can definitely do the latter, but we're using the UI to avoid the former currently 🙂

**Comment by AetherUnbound** [comment]
On the URI encoding, wow, uncovering a lot here! This was originally the solution we used, since Airflow apparently didn't support `https` as the initial scheme for HTTP connections: https://stackoverflow.com/a/69387598/3277713

It turns out though that you can also refrain from URL encoding the variable by putting the `schema` _after_ the URL, which is referenced in the original post. An example of that is also shown in the [docs you linked](https://airflow.apache.org/docs/apache-airflow-providers-http/stable/connections/http.html):

> For example:
> ```bash
> export AIRFLOW_CONN_HTTP_DEFAULT='http://username:password@servvice.com:80/https?headers=header'
> ```

Note above the trailing `/https`, which I believe directs the `HttpHook` to use `https` as the scheme instead of `http`. Now this makes things difficult! Our Slack URLs are the form of `https://hooks.slack.com/services/<secret1>/<secret2>/<secret3>`. I'd love to throw out the connection string encoding bits we have in the Docker entrypoint (not because they aren't excellent @rwidom 😉 rather, deleted code is debugged code!), but I worry doing so would lead to Airflow interpreting `/services` as the scheme to use for our Slack connections. Those connections aren't used locally, though, and I believe we define them through the UI anyway in production! And unfortunately we can't just specify `hooks.slack.com` as the HTTP connection in the environment variable since we still need to store the secrets _somewhere_.

To summarize:
- I believe this issue is probably blocking this work, so we should jump on it ASAP
- We should remove the URL-encoding steps of the entrypoint and revert to unencoded connections by default, with encoded connections where we need to specify a full path (e.g. Slack)
- Ensure that the header extras will function appropriately after all this

Does that sound okay @sarayourfriend?

**Comment by AetherUnbound** [comment]
Another approach could be that the `HttpHook` you construct explicitly overrides the [`auth_type` parameter](https://airflow.apache.org/docs/apache-airflow-providers-http/stable/_api/airflow/providers/http/hooks/http/index.html#airflow.providers.http.hooks.http.HttpHook), and we store the API token in the `password` field of the connection environment variable. `requests` doesn't seem to have its own class to handle this unfortunately: https://requests.readthedocs.io/en/latest/api/#authentication. That class just gets called with the login/password, it seems like we could potentially make a shim for it ourselves to handle `Authorization: token` use cases.

**Comment by sarayourfriend** [comment]
Originally I was passing the token as an Airflow variable to the DAG and then the HttpHook implementation just passes the Authorization header explicitly. The `run` method accepts headers :shrug: It's one faster solution.

I'd rather not get bogged down in changing the way connection strings are managed... And I still do not really understand the advantage of the connection string syntax over using explicitly configured secrets in the hook implementation class just through regular variables?

**Comment by AetherUnbound** [comment]
> I'd rather not get bogged down in changing the way connection strings are managed... And I still do not really understand the advantage of the connection string syntax over using explicitly configured secrets in the hook implementation class just through regular variables?

Honestly for HTTP connections I could see it going either way. I think for Basic Auth it works great but anything outside that (Slack & this being the perfect example) it gets messy. The advantage of having that all managed in the connection string is that you only need to know the connection ID to get everything working, Airflow will pull in the rest from the connection information. If we're going to go the Variable route though, maybe it's best to just use `requests` directly. It seems a bit unnecessary to have an entire HTTP connection that points to `github.com` 😅

**Comment by AetherUnbound** [comment]
> I'd rather not get bogged down in changing the way connection strings are managed... And I still do not really understand the advantage of the connection string syntax over using explicitly configured secrets in the hook implementation class just through regular variables?

Honestly for HTTP connections I could see it going either way. I think for Basic Auth it works great but anything outside that (Slack & this being the perfect example) it gets messy. The advantage of having that all managed in the connection string is that you only need to know the connection ID to get everything working, Airflow will pull in the rest from the connection information. If we're going to go the Variable route though, maybe it's best to just use `requests` directly. It seems a bit unnecessary to have an entire HTTP connection that points to `github.com` 😅

**Comment by sarayourfriend** [comment]
Okay, sounds good. Luckily it'll be an easy refactor :slightly_smiling_face:

**Comment by sarayourfriend** [comment]
It looks like the request to leave the comments is malformed. These logs appear when attempting to run a non-dry run of the DAG in production Airflow:

```
[2022-06-27, 00:33:17 UTC] {pr_review_reminders.py:165} INFO - Pinging @dhruvkb, @stacimc to review Add basic API load testing script
[2022-06-27, 00:33:17 UTC] {taskinstance.py:1889} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 171, in execute
    return_value = self.execute_callable()
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/airflow/operators/python.py", line 189, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/usr/local/airflow/openverse_catalog/dags/maintenance/pr_review_reminders/pr_review_reminders.py", line 167, in post_reminders
    gh.post_issue_comment(base_repo_name(pr), pr["number"], comment_body)
  File "/usr/local/airflow/openverse_catalog/dags/common/github.py", line 40, in post_issue_comment
    return self._make_request(
  File "/usr/local/airflow/openverse_catalog/dags/common/github.py", line 16, in _make_request
    response.raise_for_status()
  File "/usr/local/airflow/.local/lib/python3.10/site-packages/requests/models.py", line 960, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.github.com/repos/WordPress/openverse/issues/257/comments
```
